include_core/omrportsock.h

```diff
@@ -0,0 +1,94 @@
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#if !defined(OMRPORTSOCK_H_)
+#define OMRPORTSOCK_H_
+
+#if defined(OMR_OS_WINDOWS) && !defined(WIN32_LEAN_AND_MEAN)
+#define WIN32_LEAN_AND_MEAN
+#endif /* defined(OMR_OS_WINDOWS) && !defined(WIN32_LEAN_AND_MEAN) */
+
+#if defined(OMR_OS_WINDOWS) && defined(_WINSOCKAPI_)
+#undef _WINSOCKAPI_
+#endif /* defined(OMR_OS_WINDOWS) && defined(_WINSOCKAPI_) */
+
+#if defined(OMR_OS_WINDOWS)
+/* windows.h defined UDATA. Ignore its definition to avoid redefinition errors */
+#define UDATA UDATA_win32_
+#include <windows.h>
+#undef UDATA
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#pragma comment(lib, "Ws2_32.lib")
+#pragma comment (lib, "Mswsock.lib") 
+#pragma comment (lib, "AdvApi32.lib") 
+#else /* defined(OMR_OS_WINDOWS) */
+#include <sys/socket.h>
+#include <netinet/in.h>
+#endif /* defined(OMR_OS_WINDOWS) */
+
+#include<stdlib.h>
+#include<stdio.h>
+
+/* Common data types required for using the omr socket API */
+
+#if defined(OMR_OS_WINDOWS)
+typedef SOCKET OMRSocket;
+typedef SOCKADDR OMRSockAddr; /* for IPv4 */
+typedef SOCKADDR_IN OMRSockAddrIn;	/* for IPv4 addresses*/
+typedef SOCKADDR_IN6 OMRSockAddrIn6;  /* for IPv6 addresses*/
+typedef struct sockaddr_storage OMRSockAddrStorage; /* For IPv4 or IPv6 addresses */
+typedef struct addrinfoW OMRAddrInfo;  /* addrinfo structure – Unicode, for IPv4 or IPv6 */
+#else /* defined(OMR_OS_WINDOWS) */
+typedef int OMRSocket;
+typedef struct sockaddr OMRSockAddr;
+typedef struct sockaddr_in OMRSockAddrIn; /* for IPv4 addresses*/
+typedef struct sockaddr_in6 OMRSockAddrIn6; /* for IPv6 addresses*/
+typedef struct sockaddr_storage OMRSockAddrStorage; /* For IPv4 or IPv6 addresses */
+typedef struct addrinfo OMRAddrInfo; /* addrinfo structure for IPv4 or IPv6*/
+#endif /* defined(OMR_OS_WINDOWS) */
+
+/* Filled in using @ref omr_getaddrinfo. */
+typedef struct OMRAddrInfoNode {
+	/* Defined differently depending on the operating system.
+	 * Pointer to the first addrinfo node in listed list. 
+	 */
+	OMRAddrInfo *addrInfo;
+	/* Number of addrinfo nodes in linked list */
+	int length;
+} OMRAddrInfoNode;
+typedef OMRAddrInfoNode *omrsock_addrinfo_t;
+
+/* Used for storage of ip addresses. If we are only supporting IPv4, then we allocate smaller 
+ * storage space. 
+ */
+#if defined(IPv6_FUNCTION_SUPPORT)
```

Comment:
Good point, sorry, I forgot to check for this.

I'm not very sure but it seems like, for `j9sock.c`, `IPv6_FUNCTION_SUPPORT` is defined by `j9vm_uma_supportsIpv6`, which is derived from `uma_supportsIpv6` flag. It seems like `uma_supportsIpv6` is defined true or false for all OS and architecture, whether or not that OS/architecture supports IPv6. 

But I don't see them defined for OMR. Also, I'm not sure exactly how the `uma_supportsIpv6` works so I'm not sure if we can get the correct result just by checking its definition. So should we perhaps write a helper function, that runs before the variables are defined, to determine if IPv6 is supported (for example trying to make a IPv6 socket and see if it fails)? But I'm not sure how we can do that because the variables in `omrportsock.c` is defined during compilation (I'm not completely sure about this). Or, perhaps we can just defined the socket address to be the storage size, so we don't need to worry if it is IPv6 supported or not. We can run the same check (for example trying to make a IPv6 socket and see if it fails) when we start up the OMR Socket API because we will need a function call to start up OMR Socket API anyways when we implement for windows (to call winsock functions we have to run WSAstartup).

Reply:
Yes, it can be a dynamic decision during runtime if you want to take the IPv4 or IPv6 path. This decision can be specified by the API's consumer through an input parameter at startup. If a user doesn't specify a preference for IPv4/v6, then you can assume IPv6 to be the default with a fallback path to IPv4 in case IPv6 is not supported or does not work.

In OpenJ9, `uma_supportsIpv6` is a build flag enabled during compilation. As a designer, you need to define it for all operating systems and architectures.

Here is an example of adding a similar flag in OMR: https://github.com/eclipse/omr/pull/4086/commits/febff01882095f37f858c67c2689cb710c0b8584. In this example, the flag is named `OMR_THR_MCS_LOCKS`. `configure` file is auto-generated by invoking the `autoconf` command in the OMR repo's root directory. You will need to update the remaining five files. You can enable/disable such a flag by adding `set(OMR_THR_MCS_LOCKS [ON|OFF] CACHE BOOL "")` in a platform specific flags file (eg. `linux_x86-64.cmake`). There are two build systems `CMake` and `autoconf`. @rwy0717 do we still support both autoconf and cmake build systems?

Reply:
Unfortunately yes, you will need to support both build system.

Reply:
Sounds good. For now, I will remove this macro and try to do it dynamically. If it is needed later on during implementations, I will add the flags.